src/twisted/internet/endpoints.py

```diff
@@ -694,26 +717,43 @@ def connect(self, protocolFactory):
         connection which is established first.
         """
         wf = protocolFactory
-        d = self._nameResolution(self._host, self._port)
+        d = Deferred()
+        if self._badHostname:
+            ude = ValueError("invalid hostname: {}".format(self._hostText))
+            d.errback(ude)
+            return d
+        addresses = []
+        @provider(IResolutionReceiver)
+        class MyReceiver(object):
```

Comment:
``MyReceiver`` is a pretty crummy name, maybe something like ``EndpointReceiver``?